### ISSUE TITLE
edit_utf8()/edit_lines_utf8()/etc. methods similar to File::Slurp’s

### ISSUE THREAD

**Issue by shlomif** [dagolden/Path-Tiny#161]
One recent addition to File-Slurp that I find useful is https://metacpan.org/pod/File::Slurp#edit_file-edit_file_lines and I'd like something similar in Path-Tiny as well. What it does is encapsulate slurping the file, mutating it, and writing it back into one command.

I started working on it here:

https://github.com/shlomif/Path-Tiny/tree/edit-file-methods--a-la-File-Slurp

Thing is - are additions like these considered acceptable by the heads of Path-Tiny? Because I can complete my work, and don't want to work for naught.

Please let me know.

**Comment by jayallen** [comment]
I'm not a core dev but I love it

**Comment by shlomif** [comment]
@jayallen: thanks for the response and for your support. Let's wait for more answers.

**Comment by xdg** [comment]
Seems interesting. I wonder if putting the text in $_ and also passing it
is excessive. I wonder if passing the File object to the callback would be
useful instead.

So, provisionally, it seems narrowly scoped and useful.

**Comment by xdg** [comment]
My other thought is about whether `edit_utf8` should take an argument for the callback:

    $path->edit_utf8{ \&callback, \@callback_args )

This would give users some flexibility.

However, this doesn't follow the pattern of `visit`, which is similar.  So possibly, the signature should be:

     $path->edit_utf8{ \&callback, \%options )

And one of the options could be `args`, allowing:

     $path->edit_utf8{ \&callback, { args => [ $arg1, $arg2, ... ] } )

And that could also be added to `visit` for symmetry.

**Comment by shlomif** [comment]
@xdg : regarding also passing the $_ to the sub, that may indeed be a problem on older perls without copy-on-write. We can decide against it. About passing options to the callback , that may be added out of convenience, but is not entirely necessary due to the ability of doing this:

```
$path->edit_utf8(sub { other_cb(@_, @MORE_ARGS); });
```

But I can implement it this way if necessary.

**Comment by xdg** [comment]
I'm just brainstorming at the moment.  The problem with wrapping a callback in a sub is that you incur double sub call overhead, which sort of sucks.

Possibly, following the ::Tiny principle, we should implement *just* the callback (no options) and see what the actual demand is.

**Comment by shlomif** [comment]
@xdg : I see. Well, please let me know what the final verdict is. I'm a bit reminded of this naturally - http://bikeshed.com/ (hope you are not too offended). Perhaps I should go on in implementing the other methods in separate commits and then possibly revise them accordingly.

**Comment by xdg** [comment]
Debating what to name the function is bikeshedding.  Taking time to think
through API decisions that will last a long time is design.  Good design is
not rushed.

**Comment by shlomif** [comment]
@xdg : hi! It's been a while (over a month). Is there any verdict? Note that I started working on the changes in this branch - https://github.com/shlomif/Path-Tiny/tree/edit-file-methods--a-la-File-Slurp , but I want to get the design finalised.

**Comment by xdg** [comment]
I see you added `edit_raw`.  I think if you want to implement beyond utf8, then you need to do the same triplet as elsewhere.  I.e.:

* `edit`, `edit_utf8`, `edit_raw`
* `edit_lines`, `edit_lines_utf8`, `edit_lines_raw`

The generic `edit` and `edit_lines` should have parameters like this:
```
path("foo")->edit( $cb, $options );
path("foo")->edit_lines( $cb, $options );
```
The only valid option should be `binmode`, like slurp and spew.

For `edit_lines`,  you need to watch out for symlinks.  See the implementation of `spew`.

And you'll need tests for it all, of course.  :-)

I'd be happy to take a pull request for what I describe above.

**Comment by shlomif** [comment]
Thanks, @xdg! I think I can take it from here. I also decided to avoid passing the line/whole-text as an argument to the callback and just have it mutate $_ instead.

**Comment by xdg** [comment]
Closed via #163

**Comment by shlomif** [comment]
Thanks, @xdg !
